### PR TITLE
feat(network): broadcast stream messages

### PR DESCRIPTION
**Pull request by guy-starkware**
Adds a StreamMessage struct that can be serialized and sent over a broadcast channel.

- - -
This change is [Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/649)

### CODE REVIEW

File: crates/papyrus_protobuf/src/converters/consensus.rs [hunk 1]
```diff
@@ -107,6 +107,49 @@ impl From<Vote> for protobuf::Vote {
...
+        //     return Err(ProtobufConversionError::MissingField { field_description: "stream_id" });
+        // };
+
+        // let Some(chunk_id) = value.chunk_id else {
+        //     return Err(ProtobufConversionError::MissingField { field_description: "chunk_id" });
+        // };
+
+        // let Some(done) = value.done else {
+        //     return Err(ProtobufConversionError::MissingField { field_description: "done" });
+        // };
```
Comment (guy-starkware) [inline crates/papyrus_protobuf/src/converters/consensus.rs:129]: I am not sure why in some cases we need to check for missing fields and in other cases we can just use the values directly...

**Review by eitanm-starkware — CHANGES_REQUESTED** [review]
> Reviewed 6 of 7 files at r3, 1 of 7 files at r4, 2 of 3 files at r6, 3 of 5 files at r7, 3 of 3 files at r9, 1 of 1 files at r10, 1 of 1 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/649)* status: all files reviewed, 4 unresolved discussions (waiting on @dan-starkware, @guy-starkware, and @ShahakShama)

___
*[`-- commits` line 26 at r11](https://reviewable.io/reviews/starkware-libs/sequencer/649#-O5qtTEo3w2Npt3-3fNZ):*
merge all commits into a single commit for this pr
___
*[`crates/papyrus_protobuf/src/consensus.rs` line 46 at r11](https://reviewable.io/reviews/starkware-libs/sequencer/649#-O5qt_YW8lRjryn-Kc_o:-O5qt_YW8lRjryn-Kc_p:b13xgai) ([raw file](https://github.com/starkware-libs/sequencer/blob/43f6397b1dfe3e4fc9243f327122c34d255dead8/crates/papyrus_protobuf/src/consensus.rs#L46)):*
> ```XML
> #[derive(Debug, Default, Clone, Hash, Eq, PartialEq)]
> pub struct StreamMessage {
>     pub message: Vec<u8>,
> ```

Bytes? or did you not want the dependency?



_Code quote:_
```XML
Vec<u8>
```
___
*[`crates/papyrus_protobuf/src/converters/consensus.rs` line 129 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/649#gh-1735717306:-O5qrUR973pITtXKjfAj:bh3s0s5) ([raw file](https://github.com/starkware-libs/sequencer/blob/d7e7282f972da10fe44222ed0c5cc08776d4cb5b/crates/papyrus_protobuf/src/converters/consensus.rs#L129)):*
<details><summary><i>Previously, guy-starkware wrote…</i></summary><blockquote>

I am not sure why in some cases we need to check for missing fields and in other cases we can just use the values directly... 
</blockquote></details>

only primitive types are not declared optional in protobuf. in three cases above it should return Err as you are unwrapping non-optional types
___
*[`crates/papyrus_protobuf/src/converters/consensus.rs` line 141 at r11](https://reviewable.io/reviews/starkware-libs/sequencer/649#-O5qtwCu4XNTqR3QKo-J:-O5qtwCu4XNTqR3QKo-K:b-8bkvx0) ([raw file](https://github.com/starkware-libs/sequencer/blob/43f6397b1dfe3e4fc9243f327122c34d255dead8/crates/papyrus_protobuf/src/converters/consensus.rs#L141)):*
> ```XML
> impl From<StreamMessage> for protobuf::StreamMessage {
>     fn from(value: StreamMessage) -> Self {
>         protobuf::StreamMessage {
> ```

_Self_ is more readable



_Code quote:_
```XML
protobuf::StreamMessage
```

**Review by guy-starkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/649)* status: all files reviewed, 4 unresolved discussions (waiting on @dan-starkware, @eitanm-starkware, and @ShahakShama)

___
*[`-- commits` line 26 at r11](https://reviewable.io/reviews/starkware-libs/sequencer/649#-O5qtTEo3w2Npt3-3fNZ):*
<details><summary><i>Previously, eitanm-starkware wrote…</i></summary><blockquote>

merge all commits into a single commit for this pr
</blockquote></details>

I am planning to do a "squash and merge" when done. This is what I usually do for all PRs. Is this a good strategy in general? 
___
*[`crates/papyrus_protobuf/src/consensus.rs` line 46 at r11](https://reviewable.io/reviews/starkware-libs/sequencer/649#-O5qt_YW8lRjryn-Kc_o:-O5qxhj17114xRPv795d:bjo982n) ([raw file](https://github.com/starkware-libs/sequencer/blob/43f6397b1dfe3e4fc9243f327122c34d255dead8/crates/papyrus_protobuf/src/consensus.rs#L46)):*
<details><summary><i>Previously, eitanm-starkware wrote…</i></summary><blockquote>

Bytes? or did you not want the dependency?
</blockquote></details>

I don't know the differences well enough, so I went with something simple (ie from the standard library). Do you think there is an advantage to using `Bytes` in this case? 
___
*[`crates/papyrus_protobuf/src/converters/consensus.rs` line 129 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/649#gh-1735717306:-O5qyL7c4Sobm9jc7QXF:b-3ami89) ([raw file](https://github.com/starkware-libs/sequencer/blob/d7e7282f972da10fe44222ed0c5cc08776d4cb5b/crates/papyrus_protobuf/src/converters/consensus.rs#L129)):*
<details><summary><i>Previously, eitanm-starkware wrote…</i></summary><blockquote>

only primitive types are not declared optional in protobuf. in three cases above it should return Err as you are unwrapping non-optional types
</blockquote></details>

I am still a little confused. The primitive types arrive either with the value I put into the struct or with a default value, so there is no need to check if they exist? 
I guess we are just ignoring the possiblilty that the sender did not put values in them, or that the schema changed so the message uses new fields, etc? 
___
*[`crates/papyrus_protobuf/src/converters/consensus.rs` line 141 at r11](https://reviewable.io/reviews/starkware-libs/sequencer/649#-O5qtwCu4XNTqR3QKo-J:-O5qzKUB97bOupioJ0NB:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/43f6397b1dfe3e4fc9243f327122c34d255dead8/crates/papyrus_protobuf/src/converters/consensus.rs#L141)):*
<details><summary><i>Previously, eitanm-starkware wrote…</i></summary><blockquote>

_Self_ is more readable
</blockquote></details>

Done.

**Review by eitanm-starkware — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/649)* status: all files reviewed, 2 unresolved discussions (waiting on @dan-starkware, @guy-starkware, and @ShahakShama)

___
*[`crates/papyrus_protobuf/src/consensus.rs` line 46 at r11](https://reviewable.io/reviews/starkware-libs/sequencer/649#-O5qt_YW8lRjryn-Kc_o:-O5r6vYF31ahw_tBa0Sx:b-ldkh3l) ([raw file](https://github.com/starkware-libs/sequencer/blob/43f6397b1dfe3e4fc9243f327122c34d255dead8/crates/papyrus_protobuf/src/consensus.rs#L46)):*
<details><summary><i>Previously, guy-starkware wrote…</i></summary><blockquote>

I don't know the differences well enough, so I went with something simple (ie from the standard library). Do you think there is an advantage to using `Bytes` in this case?
</blockquote></details>

clarity
___
*[`crates/papyrus_protobuf/src/converters/consensus.rs` line 129 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/649#gh-1735717306:-O5r6ynXAQ-WXZg8XOE_:b2s6epe) ([raw file](https://github.com/starkware-libs/sequencer/blob/d7e7282f972da10fe44222ed0c5cc08776d4cb5b/crates/papyrus_protobuf/src/converters/consensus.rs#L129)):*
<details><summary><i>Previously, guy-starkware wrote…</i></summary><blockquote>

I am still a little confused. The primitive types arrive either with the value I put into the struct or with a default value, so there is no need to check if they exist? 
I guess we are just ignoring the possiblilty that the sender did not put values in them, or that the schema changed so the message uses new fields, etc?
</blockquote></details>

we discussed in person so i am resolving

**Review by eitanm-starkware — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/649#-:-O5r79fNDeRtVjIzrw6X:bnfp4nl)* status: all files reviewed, 2 unresolved discussions (waiting on @dan-starkware, @guy-starkware, and @ShahakShama)

**Review by ShahakShama — CHANGES_REQUESTED** [review]
> Reviewed 2 of 3 files at r12, 1 of 1 files at r13, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/649)* status: all files reviewed, 3 unresolved discussions (waiting on @dan-starkware, @eitanm-starkware, and @guy-starkware)

___
*[`-- commits` line 1 at r13](https://reviewable.io/reviews/starkware-libs/sequencer/649#-O5ryIhYBwhmHiRB1ZEV):*
Reminding you to squash the PR and not use rebase & merge
___
*[`crates/papyrus_protobuf/src/consensus.rs` line 46 at r11](https://reviewable.io/reviews/starkware-libs/sequencer/649#-O5qt_YW8lRjryn-Kc_o:-O5ryPAf4Evq-NnSqpPg:b-s3fm8x) ([raw file](https://github.com/starkware-libs/sequencer/blob/43f6397b1dfe3e4fc9243f327122c34d255dead8/crates/papyrus_protobuf/src/consensus.rs#L46)):*
<details><summary><i>Previously, eitanm-starkware wrote…</i></summary><blockquote>

clarity
</blockquote></details>

@eitanm-starkware this will cause a cyclic dependency. It might be an idea to move bytes to protobuf crate, but IMO we should remove this type alias and use Vec<u8> everywhere
___
*[`crates/papyrus_protobuf/src/consensus.rs` line 46 at r13](https://reviewable.io/reviews/starkware-libs/sequencer/649#-O5rynf8EgNX4BWnd7dz:-O5rynf8EgNX4BWnd7e-:b-c70ez5) ([raw file](https://github.com/starkware-libs/sequencer/blob/c8a93d0251420a45b7707bf5332c4f1217de6c96/crates/papyrus_protobuf/src/consensus.rs#L46)):*
> ```XML
> #[derive(Debug, Default, Clone, Hash, Eq, PartialEq)]
> pub struct StreamMessage {
>     pub message: Vec<u8>,
> ```

IMO this can be ConsensusMessage and not Vec<u8> since we don't have plans for using streaming in other places. @matan-starkware please verify you're ok with that. (proto should also change accordingly)

**Review by ShahakShama — CHANGES_REQUESTED** [review]
> Reviewed 1 of 3 files at r14, 2 of 2 files at r15, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/649)* status: all files reviewed, 8 unresolved discussions (waiting on @dan-starkware and @guy-starkware)

___
*[`crates/papyrus_protobuf/src/consensus.rs` line 45 at r15](https://reviewable.io/reviews/starkware-libs/sequencer/649#-O6FU9tK-_Gr6KANU7XL:-O6FU9tK-_Gr6KANU7XM:b-z60lw4) ([raw file](https://github.com/starkware-libs/sequencer/blob/5fe512cba12567d88ce561225c4acd8a13bcb976/crates/papyrus_protobuf/src/consensus.rs#L45)):*
> ```XML
> }
> #[derive(Debug, Default, Clone, Hash, Eq, PartialEq)]
> pub struct StreamMessage<T: Into<Vec<u8>> + TryFrom<Vec<u8>>> {
> ```

Require the TryFrom<Vec<u8>> to have Error = ProtobufConversionError. This will solve other issues you had
```
+ TryFrom<Vec<u8>, Error = ProtobufConversionError>>>
```
___
*[`crates/papyrus_protobuf/src/converters/consensus.rs` line 118 at r15](https://reviewable.io/reviews/starkware-libs/sequencer/649#-O6FTmfb8R3vlBLI18lx:-O6FTmfb8R3vlBLI18ly:b-ewix6e) ([raw file](https://github.com/starkware-libs/sequencer/blob/5fe512cba12567d88ce561225c4acd8a13bcb976/crates/papyrus_protobuf/src/converters/consensus.rs#L118)):*
> ```XML
> // to match ProtoBufConversionError, so I added this, as suggested by
> // https://github.com/rust-lang-deprecated/error-chain/issues/229#issuecomment-333406310
> impl From<Infallible> for ProtobufConversionError {
> ```

I don't think this is needed, and definitely not with panic statement. What is the compiler error you're getting if you remove this?
___
*[`crates/papyrus_protobuf/src/converters/consensus.rs` line 144 at r15](https://reviewable.io/reviews/starkware-libs/sequencer/649#-O6FUhwBEZua095ehWT0:-O6FUhwBEZua095ehWT1:bexatnh) ([raw file](https://github.com/starkware-libs/sequencer/blob/5fe512cba12567d88ce561225c4acd8a13bcb976/crates/papyrus_protobuf/src/converters/consensus.rs#L144)):*
> ```XML
>     Self: TryFrom<protobuf::StreamMessage>,
>     StreamMessage<T>: TryFrom<protobuf::StreamMessage>,
>     ProtobufConversionError: From<<StreamMessage<T> as TryFrom<protobuf::StreamMessage>>::Error>,
> ```

After you do the change above, remove this bound
___
*[`crates/papyrus_protobuf/src/converters/consensus.rs` line 151 at r15](https://reviewable.io/reviews/starkware-libs/sequencer/649#-O6FUdpR4X1WnWH7efX5:-O6FUdpR4X1WnWH7efX6:bdlehen) ([raw file](https://github.com/starkware-libs/sequencer/blob/5fe512cba12567d88ce561225c4acd8a13bcb976/crates/papyrus_protobuf/src/converters/consensus.rs#L151)):*
> ```XML
>         match Self::try_from(protobuf_value) {
>             Ok(value) => Ok(value),
>             Err(e) => Err(e.into()),
> ```

After you do the change above, remove the into() here
___
*[`crates/papyrus_protobuf/src/converters/consensus.rs` line 159 at r15](https://reviewable.io/reviews/starkware-libs/sequencer/649#-O6FUq3eBl1-h7MzWM--:-O6FUq3eBl1-h7MzWM-0:b-510h91) ([raw file](https://github.com/starkware-libs/sequencer/blob/5fe512cba12567d88ce561225c4acd8a13bcb976/crates/papyrus_protobuf/src/converters/consensus.rs#L159)):*
> ```XML
>     fn from(value: StreamMessage<T>) -> Self
>     where
>         T: Into<Vec<u8>>,
> ```

No need to repeat the same bound twice
___
*[`crates/papyrus_protobuf/src/converters/consensus.rs` line 172 at r15](https://reviewable.io/reviews/starkware-libs/sequencer/649#-O6FV4IwDXeDHR4F3ORe:-O6FV4IwDXeDHR4F3ORf:bvg8rsg) ([raw file](https://github.com/starkware-libs/sequencer/blob/5fe512cba12567d88ce561225c4acd8a13bcb976/crates/papyrus_protobuf/src/converters/consensus.rs#L172)):*
> ```XML
> // Can't use auto_impl_into_and_try_from_vec_u8!(StreamMessage, protobuf::StreamMessage);
> // because it doesn't seem to work with generics
> impl<T: Into<Vec<u8>> + TryFrom<Vec<u8>>> From<StreamMessage<T>> for Vec<u8> {
> ```

What about the other side (TryFrom<Vec<u8>> for StreamMessage)?
___
*[`crates/papyrus_protobuf/src/converters/consensus_test.rs` line 8 at r15](https://reviewable.io/reviews/starkware-libs/sequencer/649#-O6FVGJJ3q6kInYVycJj:-O6FVGJJ3q6kInYVycJk:b-qd971c) ([raw file](https://github.com/starkware-libs/sequencer/blob/5fe512cba12567d88ce561225c4acd8a13bcb976/crates/papyrus_protobuf/src/converters/consensus_test.rs#L8)):*
> ```XML
> // TODO(guyn): add tests for other serializable objects in consensus
> 
> impl GetTestInstance for StreamMessage<ConsensusMessage> {
> ```

Use the macro for this auto_impl_get_test_instance
___
*[`crates/papyrus_protobuf/src/converters/consensus_test.rs` line 19 at r15](https://reviewable.io/reviews/starkware-libs/sequencer/649#-O6FVSBvAq7gbUZM4tMe:-O6FVSBvAq7gbUZM4tMf:bjv6iev) ([raw file](https://github.com/starkware-libs/sequencer/blob/5fe512cba12567d88ce561225c4acd8a13bcb976/crates/papyrus_protobuf/src/converters/consensus_test.rs#L19)):*
> ```XML
> }
> 
> impl GetTestInstance for StreamMessage<Vec<u8>> {
> ```

No need to test this.

**Review by ShahakShama — APPROVED** [review]
:lgtm:

> Reviewed 1 of 3 files at r16, 2 of 2 files at r17, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/649#-:-O6Fv_JNDIEja02AX2C7:bnfp4nl)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @dan-starkware)

**Review by ShahakShama — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/649)* status: all files reviewed, 1 unresolved discussion (waiting on @dan-starkware and @guy-starkware)

___
*[`crates/papyrus_protobuf/src/proto/p2p/proto/stream_message.proto` line 10 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/649#-O6GC46sB-xV0ZfZgDdP:-O6GC46sB-xV0ZfZgDdQ:bji0p3q) ([raw file](https://github.com/starkware-libs/sequencer/blob/ea4495ede8407d7ddff375007bab2584cc29291e/crates/papyrus_protobuf/src/proto/p2p/proto/stream_message.proto#L10)):*
> ```proto
>     uint64 chunk_id = 3
>     bool done = 4
> }
> ```

remove the newline at the end of file

**Review by ShahakShama — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/649)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @dan-starkware)

**Pull request merged**